### ISSUE TITLE
Gradients are `None` when using `torch.utils.checkpoint.checkpoint`

### ISSUE THREAD

**Issue by younesbelkada** [pytorch/pytorch#112397]
### 🐛 Describe the bug

In Hugging Face transformers for some model architectures the gradients of trained models are being set to `None` when one uses gradient checkpointing. 

I am not sure if this is a bug, but I managed to reproduce the behaviour in the snippet below. 
 
```python
import torch

from torch.utils.checkpoint import checkpoint

class DummyLayer(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = torch.nn.Linear(4, 4)
        self.ln1 = torch.nn.LayerNorm(4)

        self.linear2 = torch.nn.Linear(4, 4)
        self.ln2 = torch.nn.LayerNorm(4)

    def forward(self, x):
        x = self.linear1(x)
        x = self.ln1(x)

        x = self.linear2(x)
        x = self.ln2(x)
        return x

class DummyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.emb = torch.nn.Embedding(4, 4)
        self.layers = torch.nn.ModuleList([DummyLayer() for _ in range(2)])
        self.head = torch.nn.Linear(4, 4)

        self._gradient_checkpointing_func = checkpoint

    def custom(self, module):
        def custom_forward(*inputs):
            inputs = module(inputs[0])
            return inputs
        return custom_forward
    
    def forward(self, x):
        x = self.emb(x)

        for layer in self.layers:
            x = self._gradient_checkpointing_func(self.custom(layer), x)

        return x

model = DummyModel().to(0)
model.train()

for p in model.parameters():
    p.requires_grad_(True)

assert model.training

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
dummy_input = torch.LongTensor([[0, 1, 0, 1]]).to(0)

model.train()
logits = model(dummy_input)
loss = logits.sum()

loss.backward()
optimizer.step()

for n, param in model.named_parameters():
    if param.grad is None:
        raise ValueError(f"Parameter {n} has no gradient!")
```

Note this happens also if one uses `use_reentrant=False`:

```diff
- self._gradient_checkpointing_func = checkpoint
+ self._gradient_checkpointing_func = functools.partial(checkpoint, use_reentrant=False)
```

### Versions

```
PyTorch version: 2.1.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.28

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.19.0-25-cloud-amd64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4

Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               63
Model name:          Intel(R) Xeon(R) CPU @ 2.30GHz
Stepping:            0
CPU MHz:             2299.998
BogoMIPS:            4599.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            46080K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.2
[pip3] flake8==3.7.9
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] onnx==1.12.0
[pip3] onnxconverter-common==1.13.0
[pip3] tf2onnx==1.13.0
[pip3] torch==2.1.0+cu118
[pip3] torchaudio==2.1.0+cu118
[pip3] torchdiffeq==0.2.3
[pip3] torchlibrosa==0.0.9
[pip3] torchsde==0.2.5
[pip3] torchvision==0.16.0+cu118
[pip3] triton==2.1.0
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     2.1.0+cu118              pypi_0    pypi
[conda] torchaudio                2.1.0+cu118              pypi_0    pypi
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchlibrosa              0.0.9                    pypi_0    pypi
[conda] torchsde                  0.2.5                    pypi_0    pypi
[conda] torchvision               0.16.0+cu118             pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi
```

**Comment by Chillee** [comment]
cc: @albanD

**Comment by albanD** [comment]
Hey
The error is only because the `head` module there is not used during the forward pass. So no gradient is populated for it.
Closing as expected behavior.
